### PR TITLE
New: Add config validator (refs #2179)

### PR DESCRIPTION
**Pull request by btmills**

### CODE REVIEW

File: lib/validate-options.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
+/**
+ * @fileoverview Validates rule options.
```
Comment (nzakas) [inline lib/validate-options.js]: I think this is named incorrectly. I'd call it config validator, as we may even want to add further validations, like for deprecated or removed rules.

**Comment by nzakas** [comment]
Now that I've been able to get a good look at this, I think we need to work on the interface. I think what we really want is a config validator, something like:

```js
var validator = require ('./config-validator');

// validate whole config
validator.validate(config);

// validate one rule
validator.validateRuleOptions('ruleid', [2, 'too']);
```

This way, we have one spot for any type of config validation we need.

Thoughts?

**Comment by btmills** [comment]
That makes sense. I've updated the code; what do you think now?

File: lib/config-validator.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    // Given a full schema, leave it alone
+    return options;
+}
+
+/**
+ * Gets an options schema for a rule.
+ * @param {string} id The rule's unique name.
+ * @returns {object}  vJSON Schema for the rule's options.
```
Comment (gcochard) [inline lib/config-validator.js]: typo here? should be `JSON` instead of `vJSON`.

Comment (btmills) [inline lib/config-validator.js]: Yep, fixed.

File: lib/config-validator.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+/**
+ * Validates a rule's options against its schema.
+ * @param {string} id The rule's unique name.
+ * @param {object} options The given options for the rule.
+ * @param {string} source The name of the configuration source.
+ * @returns {void}
+ */
+module.exports.validateRuleOptions = function (id, options, source) {
```
Comment (nzakas) [inline lib/config-validator.js]: Can you do:

```js
module.exports = {
   // methods
};
```

To match the other files?

File: lib/config-validator.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            message.push(
+                "\tValue \"", error.value, "\" ", error.message, ".\n"
+            );
+        });
+
+        throw new Error(message.join(""));
+    }
+};
+
+module.exports.getRuleOptionsSchema = optionsValidators;
```
Comment (nzakas) [inline lib/config-validator.js]: This looks wrong. The name is a method name but the value is an object?

**Comment by nzakas** [comment]
Better! Can you add a top-level `validate` method that can take in a whole config and validate each rule? I'd like to just add that method call everywhere instead of having multiple methods called in different places.

**Comment by btmills** [comment]
Updated. Anything else?

File: lib/config-validator.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    // Given a full schema, leave it alone
+    return schema;
+}
+
+/**
+ * Validates a rule's options against its schema.
+ * @param {string} id The rule's unique name.
+ * @param {object} options The given options for the rule.
```
Comment (hzoo) [inline lib/config-validator.js]: Should this be `{array|number}` ?

Comment (btmills) [inline lib/config-validator.js]: Good catch.

**Comment by nzakas** [comment]
Looks good. Going to hold off merging as another patch release is coming.

**Pull request merged**